File: runtime/typescript/packages/core/src/core/loader.ts
/**
 * Prompty loader — loads .prompty files into typed Prompty objects.
 *
 * Splits frontmatter (YAML) from the markdown body, resolves
 * `${protocol:value}` references (env vars, file includes),
 * and delegates to `Prompty.load()`.
 *
 * @module
 */

import { existsSync, readFileSync, realpathSync } from "node:fs";
import { dirname, extname, isAbsolute, relative, resolve } from "node:path";
import matter from "gray-matter";
import { LoadContext, SaveContext } from "../model/context.js";
import { Prompty } from "../model/agent/prompty.js";

export interface LoadOptions {
  /**
   * Additional directories that `${file:...}` references may read from.
   * The prompt file's directory is always allowed.
   */
  allowedFileRoots?: string[];
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Load a `.prompty` file and return a typed `Prompty`.
 *
 * @param path - File system path to a `.prompty` file.
 * @param options - Optional load behavior, including additional allowed file roots.
 * @returns Fully typed Prompty definition.
 */
export function load(path: string, options: LoadOptions = {}): Prompty {
  const resolved = resolve(path);
  const raw = readFileSync(resolved, "utf-8");
  return buildAgent(raw, resolved, options);
}

/**
 * Return a `SaveContext` that strips internal `__`-prefixed metadata keys.
 *
 * This is the save-side counterpart to the `LoadContext` used during
 * {@link load}. Pass it to `agent.save()`, `agent.toYaml()`, or
 * `agent.toJson()` to keep serialised output clean.
 */
export function defaultSaveContext(
  overrides?: Partial<Pick<SaveContext, "collectionFormat" | "useShorthand">>,
): SaveContext {
  return new SaveContext({
    postSave: (data: Record<string, unknown>) => {
      const meta = data["metadata"];
      if (meta && typeof meta === "object" && !Array.isArray(meta)) {
        const cleaned: Record<string, unknown> = {};
        let hasKeys = false;
        for (const [k, v] of Object.entries(meta as Record<string, unknown>)) {
          if (!k.startsWith("__")) {
            cleaned[k] = v;
            hasKeys = true;
          }
        }
        if (hasKeys) {
          data["metadata"] = cleaned;
        } else {
          delete data["metadata"];
        }
      }
      return data;
    },
    ...overrides,
  });
}

// ---------------------------------------------------------------------------
// Internal pipeline
// ---------------------------------------------------------------------------

function buildAgent(raw: string, filePath: string, options: LoadOptions): Prompty {
  // 1. Split frontmatter + body
  const { data, content } = matter(raw);

  // If there's a body (instructions), merge it in
  const frontmatter: Record<string, unknown> = data ?? {};
  if (content.trim()) {
    frontmatter.instructions = content.trim();
  }

  // 2. Load via Prompty.load() with preProcess for ${protocol:value} expansion
  const ctx = new LoadContext({
    preProcess: makePreProcess(filePath, options) as (data: Record<string, unknown>) => Record<string, unknown>,
  });
  const agent = Prompty.load(frontmatter, ctx);

  // Store source path for PromptyTool resolution (relative path lookups)
  if (!agent.metadata) {
    agent.metadata = {};
  }
  agent.metadata["__source_path"] = filePath;

  return agent;
}

// ---------------------------------------------------------------------------
// Reference resolution via preProcess
// ---------------------------------------------------------------------------

/**
 * Return a `preProcess` callback that resolves `${protocol:value}`
 * references in every dict the loader visits.
 *
 * Supported protocols:
 * - `${env:VAR_NAME}` — environment variable (required)
 * - `${env:VAR_NAME:default}` — environment variable with default
 * - `${file:relative/path}` — load file content (JSON/YAML/text)
 *
 * File references are limited to the prompt directory by default. Callers may
 * provide additional allowed roots via `allowedFileRoots`.
 */
function makePreProcess(agentFile: string, options: LoadOptions): (data: unknown) => unknown {
  const agentDir = realpathSync(dirname(agentFile));
  const allowedRoots = [
    agentDir,
    ...(options.allowedFileRoots ?? []).map((root) => canonicalizeExistingPath(resolve(root))),
  ];

  return (data: unknown): unknown => {
    if (typeof data !== "object" || data === null || Array.isArray(data)) {
      return data;
    }

    const record = data as Record<string, unknown>;
    for (const [key, value] of Object.entries(record)) {
      if (typeof value !== "string" || !value.startsWith("${") || !value.endsWith("}")) {
        continue;
      }

      const inner = value.slice(2, -1);
      const colonIdx = inner.indexOf(":");
      if (colonIdx === -1) continue;

      const protocol = inner.slice(0, colonIdx).toLowerCase();
      const val = inner.slice(colonIdx + 1);

      if (protocol === "env") {
        // Support ${env:VAR:default}
        const nextColon = val.indexOf(":");
        const varName = nextColon === -1 ? val : val.slice(0, nextColon);
        const defaultVal = nextColon === -1 ? undefined : val.slice(nextColon + 1);

        const envVal = process.env[varName];
        if (envVal !== undefined) {
          record[key] = envVal;
        } else if (defaultVal !== undefined) {
          record[key] = defaultVal;
        } else {
          throw new Error(
            `Environment variable '${varName}' not set for key '${key}'`,
          );
        }
      } else if (protocol === "file") {
        const filePath = resolveFileReference(agentDir, val, allowedRoots, key);
        record[key] = loadFileContent(filePath);
      }
    }

    return record;
  };
}

function resolveFileReference(agentDir: string, reference: string, allowedRoots: string[], key: string): string {
  const candidate = isAbsolute(reference) ? resolve(reference) : resolve(agentDir, reference);
  if (!isWithinAnyRoot(candidate, allowedRoots)) {
    throw new Error(
      `File reference '${reference}' resolves outside allowed roots for key '${key}'. Allowed roots: ${allowedRoots.join(", ")}`,
    );
  }
  if (!existsSync(candidate)) {
    throw new Error(`Referenced file '${reference}' not found for key '${key}' (resolved to ${candidate})`);
  }

  const resolved = realpathSync(candidate);
  if (!isWithinAnyRoot(resolved, allowedRoots)) {
    throw new Error(
      `File reference '${reference}' resolves outside allowed roots for key '${key}'. Allowed roots: ${allowedRoots.join(", ")}`,
    );
  }
  return resolved;
}

function isWithinAnyRoot(path: string, roots: string[]): boolean {
  return roots.some((root) => {
    const rel = relative(root, path);
    return rel === "" || (!rel.startsWith("..") && !isAbsolute(rel));
  });
}

function canonicalizeExistingPath(path: string): string {
  return existsSync(path) ? realpathSync(path) : resolve(path);
}

function loadFileContent(path: string): unknown {
  const raw = readFileSync(path, "utf-8");
  const ext = extname(path).toLowerCase();

  if (ext === ".json") {
    return JSON.parse(raw);
  }
  // For YAML we return raw string — the loader handles YAML natively
  return raw;
}

File: runtime/python/prompty/prompty/core/loader.py
"""Prompty loader — loads .prompty files into typed Prompty objects.

The loader splits frontmatter (YAML) from the markdown body, resolves
``${protocol:value}`` references (env vars, file includes), and
delegates to ``Prompty.load()`` from the generated model package.
"""

from __future__ import annotations

import json
import os
from collections.abc import Callable
from pathlib import Path
from typing import Any

import yaml

from ..model import LoadContext, Prompty, SaveContext
from .utils import load_prompty, load_prompty_async

__all__ = ["load", "load_async", "default_save_context"]


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------


def load(path: str | Path) -> Prompty:
    """Load a ``.prompty`` file and return a typed ``Prompty``.

    Parameters
    ----------
    path:
        File system path to a ``.prompty`` file.

    Returns
    -------
    Prompty
        Fully typed prompt definition.
    """
    path = Path(path).resolve()
    if not path.exists():
        raise FileNotFoundError(f"Prompty file not found: {path}")

    # 1. Split frontmatter + body
    data = load_prompty(path)

    # 2–7 shared pipeline
    return _build_agent(data, path)


async def load_async(path: str | Path) -> Prompty:
    """Async variant of :func:`load`."""
    path = Path(path).resolve()
    if not path.exists():
        raise FileNotFoundError(f"Prompty file not found: {path}")

    data = await load_prompty_async(path)
    return _build_agent(data, path)


def default_save_context(**kwargs: Any) -> SaveContext:
    """Return a ``SaveContext`` that strips internal ``__``-prefixed metadata keys.

    This is the save-side counterpart to the ``LoadContext`` used during
    :func:`load`. Pass it to ``agent.save()``, ``agent.to_yaml()``, or
    ``agent.to_json()`` to keep serialised output clean.
    """

    def _strip_internals(data: dict[str, Any]) -> dict[str, Any]:
        meta = data.get("metadata")
        if isinstance(meta, dict):
            cleaned = {k: v for k, v in meta.items() if not k.startswith("__")}
            if cleaned:
                data["metadata"] = cleaned
            else:
                data.pop("metadata", None)
        return data

    return SaveContext(post_save=_strip_internals, **kwargs)


# ---------------------------------------------------------------------------
# Internal pipeline
# ---------------------------------------------------------------------------


def _build_agent(data: dict[str, Any] | str, path: Path) -> Prompty:
    """Shared pipeline that transforms raw frontmatter dict into a Prompty."""

    # Handle body-only files (no frontmatter — parse returns a string)
    if isinstance(data, str):
        data = {"instructions": data}
    if not isinstance(data, dict):
        data = {}

    # 2. Load via Prompty.load() with pre_process for ${protocol:value} expansion
    ctx = LoadContext(pre_process=_pre_process(path))
    agent = Prompty.load(data, ctx)

    # Store source path for PromptyTool resolution (relative path lookups)
    if agent.metadata is None:
        agent.metadata = {}
    agent.metadata["__source_path"] = str(path)

    return agent


# ---------------------------------------------------------------------------
# Reference resolution via pre_process
# ---------------------------------------------------------------------------


def _pre_process(agent_file: Path) -> Callable[[Any], Any]:
    """Return a ``pre_process`` callback that resolves ``${protocol:value}``
    references in every dict the loader visits.

    Supported protocols:

    * ``${env:VAR_NAME}`` — environment variable (required)
    * ``${env:VAR_NAME:default}`` — environment variable with default
    * ``${file:relative/path}`` — load file content (JSON / YAML / text)
    """

    def process(data: Any) -> Any:
        if not isinstance(data, dict):
            return data

        for key, value in list(data.items()):
            if not isinstance(value, str) or not value.startswith("${"):
                continue

            # Must end with }
            if not value.endswith("}"):
                continue

            inner = value[2:-1]
            protocol, _, val = inner.partition(":")
            protocol = protocol.lower()

            if protocol == "env":
                # Support ${env:VAR:default}
                var_name, _, default = val.partition(":")
                env_val = os.environ.get(var_name)
                if env_val is None:
                    if default:
                        data[key] = default
                    else:
                        raise ValueError(f"Environment variable '{var_name}' not set for key '{key}'")
                else:
                    data[key] = env_val

            elif protocol == "file":
                relative_path = (agent_file.parent / val).resolve()
                if not relative_path.exists():
                    raise FileNotFoundError(
                        f"Referenced file '{val}' not found for key '{key}' (resolved to {relative_path})"
                    )
                data[key] = _load_file_content(relative_path)

        return data

    return process


def _load_file_content(path: Path) -> Any:
    """Load a file as JSON, YAML, or plain text based on its extension."""
    with open(path, encoding="utf-8") as f:
        if path.suffix == ".json":
            return json.load(f)
        elif path.suffix in (".yml", ".yaml"):
            return yaml.safe_load(f)
        else:
            return f.read()
